quickwit-search/src/root.rs

```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright (C) 2021 Quickwit Inc.
+ *
+ * Quickwit is offered under the AGPL v3.0 and as commercial software.
+ * For commercial licensing, contact us at hello@quickwit.io.
+ *
+ * AGPL:
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use tokio::task::JoinHandle;
+use tracing::*;
+
+use quickwit_metastore::{Metastore, SplitMetadata};
+use quickwit_proto::{
+    FetchDocsRequest, FetchDocsResult, Hit, LeafSearchRequest, LeafSearchResult, PartialHit,
+    SearchRequest, SearchResult,
+};
+
+use crate::client_pool::Job;
+use crate::list_relevant_splits;
+use crate::ClientPool;
+use crate::SearchClientPool;
+
+// Measure the cost associated to searching in a given split metadata.
+fn compute_split_cost(_split_metadata: &SplitMetadata) -> u32 {
+    //TODO: Have a smarter cost, by smoothing the number of docs.
+    1
+}
+
+/// Merges two partial hits that have already been sorted.
+fn merge_partial_hits(
+    partial_hits1: Vec<PartialHit>,
```

Comment:
It might not be much of an improvement with all the cloning.

You might as well stick to concatenate and sort again.

Reply:
Partial hits merging has been replaced with QuickwitCollector::merge_fruits, and this function has been modified to merge hits.
Please see https://github.com/quickwit-inc/quickwit/pull/163/commits/92eaea92de779529667d9abec0f5a7001e5718f1.